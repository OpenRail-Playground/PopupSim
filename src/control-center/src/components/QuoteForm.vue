<script setup lang="ts">
import { computed, reactive } from 'vue'
import { useVuelidate } from '@vuelidate/core'
import { required, maxLength } from '@vuelidate/validators'
import type { CoreQuote } from '@/utils/api'
import AppInput from './AppInput.vue'
import AppTextarea from './AppTextarea.vue'

const props = defineProps<{
  quote?: CoreQuote
}>()

const emit = defineEmits<{
  (e: 'submit', quote: CoreQuote): void
}>()

const maxBodyLength = 5000

const form = reactive<CoreQuote>({
  title: props.quote?.title || '',
  author: props.quote?.author || '',
  body: props.quote?.body || '',
  notes: props.quote?.notes || ''
})
const rules = computed(() => {
  return {
    title: { required, $autoDirty: true },
    author: { required, $autoDirty: true },
    body: { required, $autoDirty: true, maxLength: maxLength(maxBodyLength) }
  }
})
const v$ = useVuelidate(rules, form)

function submitForm() {
  if (!v$.value.$error) {
    emit('submit', form)
  }
}
</script>

<template>
  <form>
    <div class="row">
      <div>
        <AppInput
          :title="$t('form.title.title')"
          :placeholder="$t('form.title.placeholder')"
          id="title"
          v-model="form.title"
          :invalid="v$.title.$error"
        >
          {{ $t('form.title.error') }}
        </AppInput>
      </div>
      <div>
        <AppInput
          :title="$t('form.author.title')"
          :placeholder="$t('form.author.placeholder')"
          id="author"
          v-model="form.author"
          :invalid="v$.author.$error"
        >
          {{ $t('form.author.error') }}
        </AppInput>
      </div>
    </div>

    <AppTextarea
      :title="$t('form.body.title')"
      id="body"
      :max="maxBodyLength"
      v-model="form.body"
      :invalid="v$.body.$error"
    >
      {{ $t('form.body.error') }}
    </AppTextarea>

    <AppTextarea :title="$t('form.notes.title')" id="notes" v-model="form.notes" />

    <div class="btn-group">
      <button
        type="button"
        :disabled="v$.$invalid"
        class="elm-button"
        data-variant="primary"
        data-icon="save"
        @click="submitForm"
      >
        {{ $t('form.save') }}
      </button>
      <RouterLink
        :to="{ name: 'quotes' }"
        class="elm-button"
        data-variant="secondary"
        type="button"
        data-icon="cancel"
      >
        {{ $t('form.cancel') }}
      </RouterLink>
    </div>
  </form>
</template>

<style lang="scss" scoped>
form .row {
  margin-top: 0.5rem;
  display: flex;
  gap: 0.5rem;

  > div {
    flex-grow: 1;
    width: 100%;
  }
}
</style>
