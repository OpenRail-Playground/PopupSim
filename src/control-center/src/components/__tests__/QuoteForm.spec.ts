import { mount, shallowMount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'

import QuoteForm from '../QuoteForm.vue'
import type { CoreQuote } from '@/utils/api'
import { createI18n } from 'vue-i18n'

const quote: CoreQuote = {
  title: 'Quote 1',
  body: 'A great quote.',
  author: 'John Doe',
  notes: 'A short note'
}

describe('QuoteForm', () => {
  function createComponent(quote?: CoreQuote, mount = shallowMount) {
    const i18n = createI18n({ legacy: false })
    return mount(QuoteForm, {
      props: { quote },
      global: {
        plugins: [i18n],
        stubs: ['RouterLink']
      }
    })
  }

  test('renders the quote form', () => {
    const wrapper = createComponent(quote)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('creates an empty form with disabled submit button when no quote is passed', () => {
    const wrapper = createComponent()
    expect((wrapper.find('button[data-icon="save"]').element as HTMLInputElement).disabled).toEqual(
      true
    )
  })

  test('creates a form with pre-filled data from quote passed as prop', () => {
    const wrapper = createComponent(quote, mount)
    expect((wrapper.find('input[id="title"]').element as HTMLInputElement).value).toEqual(
      quote.title
    )
    expect((wrapper.find('input[id="author"]').element as HTMLInputElement).value).toEqual(
      quote.author
    )
    expect((wrapper.find('textarea[id="body"]').element as HTMLInputElement).value).toEqual(
      quote.body
    )
    expect((wrapper.find('textarea[id="notes"]').element as HTMLInputElement).value).toEqual(
      quote.notes
    )
  })

  test('emits the form data when submitting the form', async () => {
    const wrapper = createComponent(quote)
    await wrapper.find('button[data-icon="save"]').trigger('click')
    expect(wrapper.emitted().submit[0]).toEqual([quote])
  })
})
