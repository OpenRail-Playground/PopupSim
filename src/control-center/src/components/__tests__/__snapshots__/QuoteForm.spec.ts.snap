// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuoteForm > renders the quote form 1`] = `
<form
  data-v-a4dd84cc=""
>
  <div
    class="row"
    data-v-a4dd84cc=""
  >
    <div
      data-v-a4dd84cc=""
    >
      <app-input-stub
        data-v-a4dd84cc=""
        id="title"
        invalid="false"
        modelvalue="Quote 1"
        placeholder="form.title.placeholder"
        title="form.title.title"
      />
    </div>
    <div
      data-v-a4dd84cc=""
    >
      <app-input-stub
        data-v-a4dd84cc=""
        id="author"
        invalid="false"
        modelvalue="John Doe"
        placeholder="form.author.placeholder"
        title="form.author.title"
      />
    </div>
  </div>
  <app-textarea-stub
    data-v-a4dd84cc=""
    id="body"
    invalid="false"
    max="5000"
    modelvalue="A great quote."
    placeholder=""
    title="form.body.title"
  />
  <app-textarea-stub
    data-v-a4dd84cc=""
    id="notes"
    invalid="false"
    modelvalue="A short note"
    placeholder=""
    title="form.notes.title"
  />
  <div
    class="btn-group"
    data-v-a4dd84cc=""
  >
    <button
      class="elm-button"
      data-icon="save"
      data-v-a4dd84cc=""
      data-variant="primary"
      type="button"
    >
      form.save
    </button>
    <router-link-stub
      class="elm-button"
      data-icon="cancel"
      data-v-a4dd84cc=""
      data-variant="secondary"
      to="[object Object]"
      type="button"
    />
  </div>
</form>
`;
